subsys/fs/nvs/nvs.c

```diff
@@ -6,814 +6,764 @@
 {
-	int rc, entry_cmp;
-	int rot_cnt = 0;
-	u16_t free_up, entry_len_fl, hdr_len, slt_len;
-	struct nvs_entry entry, stored_entry;
-	off_t stored_entry_addr;
+	int rc, gc_count;
+	size_t ate_size, data_size;
+	struct nvs_ate wlk_ate;
+	u32_t wlk_addr, rd_addr;
+	u16_t sector_freespace;
+
+	ate_size = _nvs_al_size(fs, sizeof(struct nvs_ate));
+	data_size = _nvs_al_size(fs, len);
 
 
-	if ((id == NVS_ID_EMPTY) ||
-	    (id == NVS_ID_SECTOR_END) ||
-	    (len > fs->max_len) ||
+	if ((id == NVS_ID_EMPTY) || (len > (fs->sector_size - 2*ate_size)) ||
 	    ((len > 0) && (data == NULL))) {
 		return -EINVAL;
 	}
-	if ((len > 0) && (fs->free_space < fs->max_len)) {
-		return -ENOSPC;
+
+	if (fs->locked) {
+		return -EROFS;
 	}
-	if ((!len) || IS_ENABLED(CONFIG_NVS_PROTECT_FLASH)) {
-		/* find item to delete or check rewriting same data  */
-		stored_entry.id = id;
-		rc = nvs_get_last_entry(fs, &stored_entry);
-		if (!rc) {
-			entry_len_fl = _nvs_len_in_flash(fs, len);
-			if (stored_entry.len == entry_len_fl) {
-				stored_entry_addr = stored_entry.data_addr
-						    + fs->offset;
-				entry_cmp = memcmp(data,
-						   (void *) stored_entry_addr,
-						   len);
-				if (!entry_cmp) {
-					return 0;
-				}
-			}
-		} else {
+
+	/* find latest entry with same id */
+	wlk_addr = fs->ate_wra;
+	rd_addr = wlk_addr;
+
+	while (1) {
+		rd_addr = wlk_addr;
+		rc = _nvs_prev_ate(fs, &wlk_addr, &wlk_ate);
+		if (rc) {
 			return rc;
 		}
+		if ((wlk_ate.id == id) && (!_nvs_ate_crc8_check(&wlk_ate))) {
+			break;
+		}
+		if (wlk_addr == fs->ate_wra) {
+			break;
+		}
 	}
-	entry.id = id;
-	entry.len = len;
-	/* new data or data has changed */
-	while (1) {
-		rc = nvs_append(fs, &entry);
-		if (rc) {
-			if (rc == NVS_STATUS_NOSPACE) {
-				if (fs->free_space == 0) {
-					/* if we get here when there is no
-					 * free_space available this means
-					 * even deletes fails, we just
-					 * give up and say the file system
-					 * is full
-					 */
-					rc = -ENOSPC;
-					goto err;
-				}
-				if (rot_cnt == fs->sector_count) {
-					fs->free_space = 0;
-					rc = -ENOSPC;
-					goto err;
-				}
-				if (nvs_rotate(fs)) {
-					goto err;
-				}
-				rot_cnt++;
-			} else {
-				rc = -ENOSPC;
-				goto err;
+
+	if (wlk_addr != fs->ate_wra) {
+		/* previous entry found */
+		rd_addr &= 0xffff0000;
+		rd_addr += wlk_ate.offset;
+
+		if (len == 0) {
+			/* do not try to compare with empty data */
+			if (wlk_ate.len == 0) {
+				return 0;
 			}
 		} else {
-			break;
+			/* compare the data and if equal return 0 */
+			rc = _nvs_flash_block_cmp(fs, rd_addr, data, len);
+			if (rc <= 0) {
+				return rc;
+			}
 		}
+		/* data different, update free->space */
+		fs->free_space += _nvs_al_size(fs, wlk_ate.len);
+		fs->free_space += ate_size;
 	}
-	rc = nvs_flash_write(fs, entry.data_addr, data, entry.len);
-	if (rc) {
-		goto err;
+
+	if (fs->free_space < (data_size + ate_size)) {
+		return -ENOSPC;
```

Comment:
revert free space value before exit as write didn't happen - a smaller entry might fit.

Reply:
correct, wil change this